pandas/core/arrays/categorical.py

```diff
@@ -400,20 +405,34 @@ def astype(self, dtype: Dtype, copy: bool = True) -> ArrayLike:
             If copy is set to False and dtype is categorical, the original
             object is returned.
         """
-        if is_categorical_dtype(dtype):
+        if self.dtype is dtype:
+            result = self.copy() if copy else self
+
+        elif is_categorical_dtype(dtype):
             dtype = cast(Union[str, CategoricalDtype], dtype)
 
             # GH 10696/18593
             dtype = self.dtype.update_dtype(dtype)
             self = self.copy() if copy else self
-            if dtype == self.dtype:
-                return self
-            return self._set_dtype(dtype)
-        if is_extension_array_dtype(dtype):
-            return array(self, dtype=dtype, copy=copy)
-        if is_integer_dtype(dtype) and self.isna().any():
+            result = self._set_dtype(dtype)
+
+        elif is_extension_array_dtype(dtype):
+            result = array(self, dtype=dtype, copy=copy)
+
+        elif is_integer_dtype(dtype) and self.isna().any():
             raise ValueError("Cannot convert float NaN to integer")
-        return np.array(self, dtype=dtype, copy=copy)
+
+        elif len(self.codes) == 0 or len(self.categories) == 0:
+            result = np.array(self, dtype=dtype, copy=copy)
+
+        else:
+            # GH8628 (PERF): astype category codes instead of astyping array
+            new_categories = np.append(
+                self.categories.astype(dtype), [na_value_for_dtype(dtype)]
```

Comment:
we *never* should add the na_value into the categories themselves. this should have broken things. (codes indicates a -1) for the missing value.

Reply:
I thought this isn't the right way to do it. Though this actually this fixes the tests because in the slice `new_categories[self.codes]` we need `-1` to pick out the na_value and with this it does

(This is why the previous variant I had wasn't going green)

Reply:
> we _never_ should add the na_value into the categories themselves. this should have broken things. (codes indicates a -1) for the missing value.

Rewrote this using `take_1d` like we do in some other methods. Passes all tests modulo the error messages (commented below)